### PR TITLE
Add sage

### PR DESCRIPTION
**Pull request by isuruf**
~~@embray, @saraedum we need to figure out how to pin the version of these packages.~~

### CODE REVIEW

File: recipes/sage/build.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
...
+# patch, pkgconf, gcc, git
+
+for pkg in alabaster appnope arb babel backports_abc backports_shutil_get_terminal_size backports_ssl_match_hostname boost_cropped brial bzip2 cddlib cephes cliquer combinatorial_designs configparser conway_polynomials cvxopt cycler dateutil docutils ecl eclib ecm elliptic_curves entrypoints fflas_ffpack flask flask_babel flask_openid flint flintqs fpylll freetype functools32 gap gc gf2x gfan giac givaro glpk gmp graphs gsl iconv imagesize iml ipykernel ipython_genutils ipywidgets itsdangerous jsonschema jupyter_client jupyter_core lcalc libfplll libgap libgd libpng linbox lrcalc m4ri m4rie matplotlib mistune mpc mpfi mpfr mpmath nauty nbconvert nbformat ncurses networkx notebook ntl openblas palp pari pari_galdata pari_seadata_small pathpy pillow planarity polytopes_db ppl prompt_toolkit pycrypto pynac pyparsing python_openid pytz pyzmq r ratpoints readline rubiks rw scipy setuptools_scm singledispatch singular snowballstemmer speaklater sphinx sqlite symmetrica sympy terminado tornado vcversioner werkzeug widgetsnbextension zeromq zlib zn_poly; do
```
Comment (saraedum) [inline recipes/sage/build.sh]: Do you need this so `is_package_installed` detects them? If so, there is a patch to fix this…
https://trac.sagemath.org/ticket/20382

File: recipes/sage/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    - appnope  # [osx]
+    - arb
+    - babel
+    - backports.shutil_get_terminal_size
+    - backports_abc
+    - bdw-gc
+    - blas 1.1 openblas
+    - boost-cpp 1.63.*
+    - brial
+    - bzip2 1.0.*
```
Comment (saraedum) [inline recipes/sage/meta.yaml]: Do we need this?

File: recipes/sage/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    - flask-babel
+    - flask-openid
+    - flintqs
+    - fplll
+    - fpylll
+    - freetype 2.7|2.7.*
+    - functools32
+    - future
+    - gap
+    #- gcc
```
Comment (saraedum) [inline recipes/sage/meta.yaml]: I think this is also only a build dependency.

File: recipes/sage/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    - gf2x
+    - gfan
+    - giac
+    # is this necessary?
+    #- git
```
Comment (saraedum) [inline recipes/sage/meta.yaml]: Not necessary. Only for developers.

File: recipes/sage/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    - planarity
+    - ppl
+    - prompt_toolkit
+    - psutil
+    - ptyprocess
+    - pycrypto
+    - pygments
+    - pynac
+    - pyparsing
+    - python
```
Comment (saraedum) [inline recipes/sage/meta.yaml]: Are runtime dependencies automatically build time dependencies? If so, then this is a duplicate. Also, this must be python 2.7. Not sure if that's automatically the case here already.

**Comment by saraedum** [comment]
@isuruf What do you mean by "how to pin the version of these packages"?

File: recipes/sage/build.sh (hunk 1 above)
Comment (isuruf) [inline recipes/sage/build.sh]: Yes. Thanks. I'll add this patch to sagelib.

File: recipes/sage/meta.yaml (hunk 2 above)
Comment (isuruf) [inline recipes/sage/meta.yaml]: Ah. will remove

File: recipes/sage/meta.yaml (hunk 5 above)
Comment (isuruf) [inline recipes/sage/meta.yaml]: No, they have to be added separately. `conda` handles packages with features specially and will use the version that it was built with. Also `skip: true  # [py>27]` statement skips other python versions

**Comment by isuruf** [comment]
For example, since sage was built with mpfr 3, it needs to be pinned to `mpfr 3.*`, which means you can have any mpfr 3.* version. This is so that when mpfr changes the `so` name for `mpfr 4.*`, `mpfr` cannot be upgraded when sage is in the conda environment, unless there is  sage version compiled with `mpfr 4.*`.
These pinnings should have happened at `sagelib`, but it's hard to work with a package that takes 1 hour to build. That's why I think pinnings should happen in `sage`

**Comment by saraedum** [comment]
@isuruf: I am not sure that I understand the comment about pinning in sagelib. What goes wrong if we pin sagelib? Also, note that builds are relatively fast with ccache. (If you do not have to build dependencies.)

File: recipes/sage/meta.yaml (hunk 2 above)
Comment (embray) [inline recipes/sage/meta.yaml]: 👍 

**Comment by isuruf** [comment]
I created an issue on sagelib for pinning dependencies, https://github.com/conda-forge/sagelib-feedstock/issues/6

**Comment by isuruf** [comment]
@ocefpaf, ready for review.

**Comment by isuruf** [comment]
@ocefpaf, do you know why Linux builds were skipped? When I try to rerender, it says
```
Package missing in current linux-64 channels: 
  - r-base
```

**Comment by ocefpaf** [comment]
I made a comment on your PR in the feedstock. You need to add some "skips" to make it work.

**Pull request merged**